### PR TITLE
Auth setup

### PR DESCRIPTION
**Pull request by nully0x**
In this PR:
- it addresses issues #84 , #85 , #86 and #88 
- setup up backend auth as required
- add some packages to support auth eg: jwt

### CODE REVIEW

File: api/app/utils/auth.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    return response.json();
+  }
+  
+  // Generate JWT with user information
+  export function generateJwtToken(user: any) {
+    const secretKey = `process.env.JWT_SECRET_KEY`;
+    const token = jwt.sign(
+      { userId: user.id, username: user.githubUsername, email: user.email, permissions: user.permissions, githubAuthToken: user.authToken },
+      secretKey,
+      { expiresIn: "1h" }
```
Comment (Emmanuel-Develops) [inline api/app/utils/auth.ts:26]: We have an hardcoded expiry time here, is that fine?

Also, I think it should be synced to the github OAuth token. That is next-auth is in control of new OAuth tokens which in turn controls generation of JWT. 
Handling refresh token with next-auth might get convoluted

File: api/app/controllers/user.controller.ts [hunk 2]
```diff
@@ -42,6 +53,53 @@ export async function create(req: Request, res: Response) {
...
+export async function signIn(req: Request, res: Response) {
+  try {
+    // Validate request
+    if (!req.body.oauthToken) {
+      return res.status(400).send({
+        message: "OAuth token can not be empty!",
+      });
+    }
+
+    const { oauthToken } = req.body;
```
Comment (Emmanuel-Develops) [inline api/app/controllers/user.controller.ts:66]: In signup it is request.body.githubAuthToken, here it is oauthToken.
For standardization, let's use the same name. Is there a specific reson for the segregation?

File: api/app/utils/auth.ts (hunk 1 above)
Comment (nully0x) [reply api/app/utils/auth.ts:26]: The time is for testing, we can move it to a constant in the last PR. 

On the second point are you saying we should remove the expiresIn on the backend or should we sync the time the one provided in the next OAuth?

File: api/app/controllers/user.controller.ts (hunk 2 above)
Comment (nully0x) [reply api/app/controllers/user.controller.ts:66]: Sure can change that. @Extheoisah any opinion

File: api/app/utils/auth.ts (hunk 1 above)
Comment (nully0x) [reply api/app/utils/auth.ts:26]: For more context, we can get user permission level from the githubAuthToken data. So we can encode that with the jwt and the githubAuthToken, or we can do that with middleware on the routes for all permission levels.

File: api/app/controllers/user.controller.ts (hunk 2 above)
Comment (Extheoisah) [reply api/app/controllers/user.controller.ts:66]: I agree.

File: api/app/controllers/user.controller.ts [hunk 3]
```diff
@@ -4,32 +4,43 @@ import { Review, Transcript, User, Wallet } from "../db/models";
...
-    githubUsername: req.body.username,
-    permissions: req.body.permissions,
-    email: req.body.email,
-  };
+    // Create a User
+    const userDetails = {
+      githubUsername: req.body.username,
+      permissions: req.body.permissions,
+      email: req.body.email,
+      authToken: req.body.githubAuthToken,
```
Comment (Extheoisah) [inline api/app/controllers/user.controller.ts:25]: This authToken field has been designated for alby. Remember we plan to change the name to albyAuth token.

File: api/app/utils/auth.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import jwt from "jsonwebtoken";
+
+// Verify the OAuth token with GitHub API
+export async function verifyGitHubToken(token: string) {
+    // Use your GitHub API endpoint and headers
+    const response = await fetch("https://api.github.com/user", {
+      headers: {
+        Authorization: `Bearer ${process.env.GITHUB_OAUTH_TOKEN}`,
```
[the remark this answers is not in the document]
Comment (nully0x) [reply api/app/utils/auth.ts:8]: No, the token here is to allow the API to be able to check for authenticated user, this is generated from  the GitHub setting not the one from next auth

[the remark this answers is not in the document]
Comment (Extheoisah) [reply api/app/utils/auth.ts:8]: So who has the token persisted in .env?

[the remark this answers is not in the document]
Comment (Extheoisah) [reply api/app/utils/auth.ts:8]: This token is supposed to be the user token for us to verify

**Pull request closed without merging**